spec/support/system_helpers.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+module SystemHelpers
+  def sign_in(user)
+    visit "/session/#{user.encoded_username}/become"
+  end
+
+  def setup_system_test
+    SiteSetting.login_required = false
+    SiteSetting.content_security_policy = false
+    SiteSetting.force_hostname = "#{Capybara.server_host}:#{Capybara.server_port}"
+    SiteSetting.external_system_avatars_enabled = false
+  end
+
+  def wait_for_record(&block)
```

Comment:
I don't like having this really...but there are times where we eagerly fill in the frontend CSS/components when saving something, even when the AJAX call has not been completed and filled in through the layers of the database, so sometimes we do need to do a little extra waiting to get the record from postgres. See the bookmark spec for usage.

Reply:
Yeah this seems like a necessary evil. The PageObject matchers automatically 'keep trying' until they succeed, but there doesn't seem to be any built-in way to do the same for non-page-object assertions.

I wonder if we should make this helper more generic? There's nothing about it which makes it 'record' specific.

For our internal haproxy manager, I wrote an rspec helper like this:

```ruby
def try_until_success(timeout: 2, frequency: 0.01)
  iterations ||= 0
  yield
rescue RSpec::Expectations::ExpectationNotMetError
  if iterations >= timeout / frequency
    raise
  end
  iterations = iterations + 1
  sleep frequency
  retry
end
```

And then it can be used like

```ruby
it "works eventually" do
  expect("blah").to eq("blah")
  
  # something async

  try_until_success do
    expect(Bookmark.find_by(bookmarkable: topic, user: user)).not_to eq(nil)
  end
end
```

Reply:
Yes I think that is better, thanks for the recommendation, will change :+1:

Reply:
Actually not sure about this part:

```
iterations >= timeout / frequency
```

This could cause, for this example, 200 DB calls before failing if the record doesn't exist. Shouldn't we just go by time passed instead with the timeout? Also I think increasing sleep time by frequency after each try for a backoff would be good...something like this:

```ruby
def try_until_success(timeout: 2, frequency: 0.01)
  start ||= Time.zone.now
  backoff ||= frequency
  yield
rescue RSpec::Expectations::ExpectationNotMetError
  if Time.zone.now >= start + timeout.seconds
    raise
  end
  sleep backoff
  backoff += frequency
  retry
end
```

Maybe it's not a concern to be hitting the DB that much? Thoughts?

Reply:
I don't think it's a massive concern - the Ruby app and Database will be running in a separate process to the web browser.

Adding a little backoff makes sense, but I think we should cap it. In your '200 iterations' example, that would mean 2 seconds between db queries on the final run, which seems a little slow. We'll also need to update the maths so that `timeout` still works correct even with the backoff. I guess we take the `Time.now` at the beginning and compare it after each `rescue`.